[CI] AutoCreateDataStreamIT testCanAutoCreateDataStreamWhenAutoCreateIndexDisabled failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/vmpraqrlh23ee/tests/:x-pack:plugin:data-streams:qa:rest:yamlRestTest/org.elasticsearch.xpack.datastreams.AutoCreateDataStreamIT/testCanAutoCreateDataStreamWhenAutoCreateIndexDisabled

**Reproduction line:**
`./gradlew ':x-pack:plugin:data-streams:qa:rest:yamlRestTest' --tests "org.elasticsearch.xpack.datastreams.AutoCreateDataStreamIT.testCanAutoCreateDataStreamWhenAutoCreateIndexDisabled" -Dtests.seed=4CB80FE696899AE5 -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=ar-DZ -Dtests.timezone=America/Argentina/Salta -Druntime.java=11`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.datastreams.AutoCreateDataStreamIT&tests.test=testCanAutoCreateDataStreamWhenAutoCreateIndexDisabled

**Failure excerpt:**
```
java.net.ConnectException: Connection refused

  at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:885)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:283)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:286)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270)
  at org.elasticsearch.test.rest.ESRestTestCase.initClient(ESRestTestCase.java:213)
  at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(null:-1)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:566)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:980)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:834)

  Caused by: java.net.ConnectException: Connection refused

    at sun.nio.ch.SocketChannelImpl.checkConnect(SocketChannelImpl.java:-2)
    at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvent(DefaultConnectingIOReactor.java:174)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:148)
    at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
    at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
    at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
    at java.lang.Thread.run(Thread.java:834)

```

The node crashed because of an assertion error (which prevents tests from executing):

```
java.lang.AssertionError: expected [index.mode] to be private but it was not |  
-- | --
  | »  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1062) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1052) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1387) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1308) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1303) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:519) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService$4.execute(MetadataIndexTemplateService.java:424) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:744) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:360) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:223) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:160) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:259) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:222) ~[elasticsearch-8.0.0.jar:8.0.0] |  
  | »  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?] |  
  | »  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?] |  
  | »  	at java.lang.Thread.run(Thread.java:834) [?:?]
```

This doesn't seem to be a data stream related error. @nik9000 @imotov Is this a failure that tsdb team should look into?

I'll look.

On Tue, Oct 5, 2021, 06:17 Martijn van Groningen ***@***.***>
wrote:

> The node crashed because of an assertion error (which prevents tests from
> executing):
>
> java.lang.AssertionError: expected [index.mode] to be private but it was not |
>
> -- | --
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1062) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1052) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1387) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1308) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1303) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:519) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService$4.execute(MetadataIndexTemplateService.java:424) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:744) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:360) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:223) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:160) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:259) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:222) ~[elasticsearch-8.0.0.jar:8.0.0] |
>
>   | »  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?] |
>
>   | »  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?] |
>
>   | »  	at java.lang.Thread.run(Thread.java:834) [?:?]
>
>
> This doesn't seem to be a data stream related error. @nik9000
> <https://github.com/nik9000> @imotov <https://github.com/imotov> Is this
> a failure that tsdb team should look into?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/issues/78677#issuecomment-934271031>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIQOJ5P44OFRK3NMI2DUFLGEFANCNFSM5FLLFKVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Here's the error:
```
»    ↓ errors and warnings from /dev/shm/elastic+elasticsearch+master+periodic+release-tests/x-pack/plugin/data-streams/qa/rest/build/testclusters/yamlRestTest-0/logs/yamlRestTest.log ↓
» [2021-10-05T08:08:57,314][WARN ][o.e.d.FileBasedSeedHostsProvider] [yamlRestTest-0] expected, but did not find, a dynamic hosts list at [/dev/shm/elastic+elasticsearch+master+periodic+release-tests/x-pack/plugin/data-streams/qa/rest/build/testclusters/yamlRestTest-0/config/unicast_hosts.txt]
» [2021-10-05T08:09:10,750][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [yamlRestTest-0] fatal error in thread [elasticsearch[yamlRestTest-0][masterService#updateTask][T#1]], exiting
»  java.lang.AssertionError: expected [index.mode] to be private but it was not
»  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1062) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1052) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1387) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1308) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:1303) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:519) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService$4.execute(MetadataIndexTemplateService.java:424) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:744) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:360) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:223) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:160) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:259) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:222) ~[elasticsearch-8.0.0.jar:8.0.0]
»  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
»  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
»  	at java.lang.Thread.run(Thread.java:834) [?:?]
```

So there are two problems:
1. We don't register the setting for the tests. I'll open a PR to fix that.
2. When you send an unsupported setting for data streams we trip an assertion. This seems bad.

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.put_settings/10_basic.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.put_settings/10_basic.yml
@@ -130,6 +130,7 @@ setup:
 ---
 put unsupported setting:
   - do:
+      catch: bad_request
       indices.put_settings:
         index: test-index
         body:

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.put_settings/10_basic.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.put_settings/10_basic.yml
@@ -126,3 +126,11 @@ setup:
      test-index.settings.index.query_string.lenient: "true"
   - match:
      test-index.settings.index.translog.durability: "async"
+
+---
+put unsupported setting:
+  - do:
+      indices.put_settings:
+        index: test-index
+        body:
+          garbage: garbage

--- a/x-pack/plugin/data-streams/qa/rest/build.gradle
+++ b/x-pack/plugin/data-streams/qa/rest/build.gradle
@@ -21,6 +21,9 @@ testClusters.configureEach {
   // disable ILM history, since it disturbs tests using _all
   setting 'indices.lifecycle.history_index_enabled', 'false'
   setting 'xpack.security.enabled', 'false'
+  if (BuildParams.isSnapshotBuild() == false) {
+    systemProperty 'es.index_mode_feature_flag_registered', 'true'
+  }
 }
 if (BuildParams.inFipsJvm){
   // These fail in CI but only when run as part of checkPart2 and not individually.

--- a/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/data_stream/160_unsupported_setting.yml
+++ b/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/data_stream/160_unsupported_setting.yml
@@ -1,0 +1,32 @@
+bad setting fails:
+  - skip:
+      version: all
+      reason: https://github.com/elastic/elasticsearch/issues/78677
+      features: allowed_warnings
+
+  - do:
+      allowed_warnings:
+        - "index template [tsdbds-template1] has index patterns [simple-data-stream1] matching patterns from existing older templates [global] with patterns (global => [*]); this template [my-template1] will take precedence during new index creation"
+      indices.put_index_template:
+        name: my-template1
+        body:
+          index_patterns: [gargage*]
+          data_stream: {}
+          template:
+            settings:
+              index:
+                number_of_replicas: 0
+                number_of_shards: 2
+                garbage: garbage
+            mappings:
+              properties:
+                "@timestamp":
+                  type: date
+
+  - do:
+      bulk:
+        refresh: true
+        index: garbage
+        body:
+          - '{"create": {}}'
+          - '{"@timestamp": "2021-04-28T18:50:04.467Z"}'

